ext/extralite/common.c

```diff
@@ -22,49 +23,62 @@ static inline VALUE get_column_value(sqlite3_stmt *stmt, int col, int type) {
   return Qnil;
 }
 
-void bind_parameter_value(sqlite3_stmt *stmt, int pos, VALUE value);
+static void bind_parameter_value(sqlite3_stmt *stmt, int pos, VALUE value, VALUE unhandled_parameter_proc, bool nested);
 
-static inline void bind_key_value(sqlite3_stmt *stmt, VALUE k, VALUE v) {
+static inline void bind_key_value(sqlite3_stmt *stmt, VALUE k, VALUE v, VALUE unhandled_parameter_proc) {
   switch (TYPE(k)) {
     case T_FIXNUM:
-      bind_parameter_value(stmt, FIX2INT(k), v);
+      bind_parameter_value(stmt, FIX2INT(k), v, unhandled_parameter_proc, true);
       break;
     case T_SYMBOL:
       k = rb_funcall(k, ID_to_s, 0);
     case T_STRING:
       if (RSTRING_PTR(k)[0] != ':') k = rb_str_plus(rb_str_new2(":"), k);
       int pos = sqlite3_bind_parameter_index(stmt, StringValuePtr(k));
-      bind_parameter_value(stmt, pos, v);
+      bind_parameter_value(stmt, pos, v, unhandled_parameter_proc, true);
       break;
     default:
       rb_raise(cParameterError, "Cannot bind parameter with a key of type %"PRIsVALUE"",
         rb_class_name(rb_obj_class(k)));
   }
 }
 
-void bind_hash_parameter_values(sqlite3_stmt *stmt, VALUE hash) {
+static inline void bind_hash_parameter_values(sqlite3_stmt *stmt, VALUE hash, VALUE unhandled_parameter_proc) {
   VALUE keys = rb_funcall(hash, ID_keys, 0);
   long len = RARRAY_LEN(keys);
   for (long i = 0; i < len; i++) {
     VALUE k = RARRAY_AREF(keys, i);
     VALUE v = rb_hash_aref(hash, k);
-    bind_key_value(stmt, k, v);
+    bind_key_value(stmt, k, v, unhandled_parameter_proc);
   }
   RB_GC_GUARD(keys);
 }
 
-void bind_struct_parameter_values(sqlite3_stmt *stmt, VALUE struct_obj) {
+static inline void bind_struct_parameter_values(sqlite3_stmt *stmt, VALUE struct_obj, VALUE unhandled_parameter_proc) {
   VALUE members = rb_struct_members(struct_obj);
   for (long i = 0; i < RSTRUCT_LEN(struct_obj); i++) {
     VALUE k = rb_ary_entry(members, i);
     VALUE v = RSTRUCT_GET(struct_obj, i);
-    bind_key_value(stmt, k, v);
+    bind_key_value(stmt, k, v, unhandled_parameter_proc);
   }
   RB_GC_GUARD(members);
 }
 
-inline void bind_parameter_value(sqlite3_stmt *stmt, int pos, VALUE value) {
-  switch (TYPE(value)) {
+static void bind_parameter_value(sqlite3_stmt *stmt, int pos, VALUE value, VALUE unhandled_parameter_proc, bool nested) {
+  int type = TYPE(value);
+
+  if (!nested) {
+    switch (type) {
+      case T_HASH:
+        bind_hash_parameter_values(stmt, value, unhandled_parameter_proc);
+        return;
+      case T_STRUCT:
+        bind_struct_parameter_values(stmt, value, unhandled_parameter_proc);
```

Comment:
I haven't looked at the implementation of `to_h` for structs yet, but I assume it will generate a new hash object. Having the separate method for structs avoids that.